### DISCUSSION TITLE
Persistent $TLR and Probe vs Toolchange events

### DISCUSSION THREAD

**Discussion by Dietz0r** [grblHAL/core#200] [Q&A]
Hey there.
I'm currently working on a Datronstyle retractable 3D Stylus Probe for my z-axis and a few problems have come up right that i wanted to discuss.

Since my probe/spindle relationship will be a fixed offset when the probe is installed on it's retractable mount instead of the spindlecollet i would like to save a persistent $TLR  so that my empty spindlecollet is the 0 point i can relate the stylus tip to.  Since no installed tool could be lower than the empty collet this would give true stickout measurements for subsequent toolchanges. Could it be possible to give that option or achieve it through different means?

Also since the probe is retractable and sitting lower than the installed tools when deployed i was wondering if there is a way to differentiate from true probing events like a G38 call from Toolchanges like M6 because i'd like to try and write a plugin to inject commands (offsett movements and M codes for probe deploy) infront of some of them but obviously dont want the probe to deploy on a toolchange.

thanks in advance :)

**Comment by terjeio** [comment]
>  save a persistent $TLR

Plugin code can [add settings](https://github.com/grblHAL/Templates/blob/master/my_plugin/settings/my_plugin.c), you may also designate an offset such as G59.2 for that.

>  if there is a way to differentiate from true probing events like a G38 call from Toolchanges like M6

Yes, toolchanges are passed over `hal.tool.select` and `hal.tool.change`, you can insert your code by saving away the pointer(s) and change it/them to your own function(s) as needed. You can also subscribe to `grbl.on_state_change` in order to be informed about all state changes, tool change in progress beeing one of them (this will not work with an ATC). Probing events are passed over `grbl.on_probe_start`, `grbl.on_probe_fixture` and `grbl.on_probe_completed`.

**Reply by Dietz0r** [reply]
thank you, i'll dig into that for a bit
